src/Workspaces/VisualBasic/Portable/Editing/VisualBasicImportAdder.vb

```diff
@@ -57,91 +59,139 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Editing
+            ''' </summary>
+            Private ReadOnly _extensionMethods As MultiDictionary(Of String, INamespaceSymbol) =
+                New MultiDictionary(Of String, INamespaceSymbol)()
+
+            Private ReadOnly _conflictNamespaces As HashSet(Of INamespaceSymbol)
+
+            ''' <summary>
+            ''' Track if we're in an anonymous method or not.  If so, because of how the language binds lambdas and
+            ''' overloads, we'll assume any method access we see inside (instance or otherwise) could end up conflicting
+            ''' with an extension method we might pull in.
+            ''' </summary>
+            Private _inAnonymousMethod As Boolean
+
+            Public Sub New(
+                    model As SemanticModel,
+                    namespaceSymbols As ImmutableArray(Of INamespaceSymbol),
+                    conflictNamespaces As HashSet(Of INamespaceSymbol),
+                    cancellationToken As CancellationToken)
+                MyBase.New(SyntaxWalkerDepth.StructuredTrivia)
                 _model = model
-                _workspace = workspace
                 _cancellationToken = cancellationToken
-                _namespaceMembers = New HashSet(Of String)(namespaceMembers.[Select](Function(x) x.Name), CaseInsensitiveComparison.Comparer)
-                _extensionMethods = New HashSet(Of String)(extensionMethods.[Select](Function(x) x.Name), CaseInsensitiveComparison.Comparer)
+                _conflictNamespaces = conflictNamespaces
+
+                For Each ns In namespaceSymbols
+                    For Each typeOrNamespace In ns.GetMembers()
+                        _namespaceMembers.Add((typeOrNamespace.Name, typeOrNamespace.GetArity()), ns)
+
+                        Dim type = TryCast(typeOrNamespace, INamedTypeSymbol)
+                        If type?.MightContainExtensionMethods Then
+                            For Each member In type.GetMembers()
+                                Dim method = TryCast(member, IMethodSymbol)
+                                If method?.IsExtensionMethod Then
+                                    _extensionMethods.Add(method.Name, ns)
+                                End If
+                            Next
+                        End If
+                    Next
+                Next
+            End Sub
+
+            Public Overrides Sub VisitMultiLineLambdaExpression(node As MultiLineLambdaExpressionSyntax)
+
+                ' lambdas are interesting.  Say you have
+                '
+                '      Goo(sub (x) x.M())
+                '
+                '      sub Goo(act as Action(of C))
+                '      sub Goo(act as Action(of integer))
+                '
+                '      class C : public sub M()
+                '
+                ' This Is legal code where the lambda body Is calling the instance method.  However, if we introduce a
```

Comment:
I'm loving all the capital "Is" here where clearly the VB formatter got it somehow. :-)